{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import torch\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "import scDART.utils as utils\n",
    "import scDART.TI as ti\n",
    "import scDART\n",
    "\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Train model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Dataset\n",
      "running time(sec): 1.30973482131958\n",
      "running time(sec): 1.3091225624084473\n",
      "running time(sec): 1.273963212966919\n",
      "running time(sec): 0.5780971050262451\n",
      "running time(sec): 0.5638875961303711\n",
      "running time(sec): 0.5683960914611816\n",
      "epoch:  0\n",
      "\t mmd loss: 0.027\n",
      "\t ATAC dist loss: 0.657\n",
      "\t RNA dist loss: 0.342\n",
      "\t gene activity loss: 5077.949\n",
      "\t anchor matching loss: 0.000\n",
      "epoch:  100\n",
      "\t mmd loss: 0.024\n",
      "\t ATAC dist loss: 0.025\n",
      "\t RNA dist loss: 0.024\n",
      "\t gene activity loss: 25.837\n",
      "\t anchor matching loss: 0.000\n",
      "epoch:  200\n",
      "\t mmd loss: 0.024\n",
      "\t ATAC dist loss: 0.018\n",
      "\t RNA dist loss: 0.018\n",
      "\t gene activity loss: 16.632\n",
      "\t anchor matching loss: 0.000\n",
      "epoch:  300\n",
      "\t mmd loss: 0.023\n",
      "\t ATAC dist loss: 0.016\n",
      "\t RNA dist loss: 0.018\n",
      "\t gene activity loss: 12.242\n",
      "\t anchor matching loss: 0.000\n",
      "epoch:  400\n",
      "\t mmd loss: 0.020\n",
      "\t ATAC dist loss: 0.015\n",
      "\t RNA dist loss: 0.017\n",
      "\t gene activity loss: 9.554\n",
      "\t anchor matching loss: 0.000\n"
     ]
    }
   ],
   "source": [
    "# all in one\n",
    "seeds = [0]\n",
    "latent_dim = 4\n",
    "learning_rate = 3e-4\n",
    "n_epochs = 500\n",
    "use_anchor = False\n",
    "reg_d = 1\n",
    "reg_g = 1\n",
    "reg_mmd = 1\n",
    "ts = [30, 50, 70]\n",
    "use_potential = True\n",
    "\n",
    "counts_rna = pd.read_csv(\"../data/Endo/counts_rna.csv\", index_col = 0)\n",
    "counts_atac = pd.read_csv(\"../data/Endo/counts_atac.csv\", index_col = 0)\n",
    "label_rna = pd.read_csv(\"../data/Endo/anno_rna.txt\", header = None)\n",
    "label_atac = pd.read_csv(\"../data/Endo/anno_atac.txt\", header = None)\n",
    "coarse_reg = pd.read_csv(\"../data/Endo/region2gene.csv\", sep = \",\", index_col = 0).values\n",
    "\n",
    "\n",
    "scDART_op = scDART.scDART(n_epochs = n_epochs, latent_dim = latent_dim, \\\n",
    "        ts = ts, use_anchor = use_anchor, use_potential = use_potential, k = 10, \\\n",
    "        reg_d = 1, reg_g = 1, reg_mmd = 1, l_dist_type = 'kl', seed = seeds[0],\\\n",
    "        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu'))\n",
    "\n",
    "scDART_op = scDART_op.fit(rna_count = counts_rna.values, atac_count = counts_atac.values, reg = coarse_reg, rna_anchor = None, atac_anchor = None)\n",
    "z_rna, z_atac = scDART_op.transform(rna_count = counts_rna.values, atac_count = counts_atac.values, rna_anchor = None, atac_anchor = None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import scDART.diffusion_dist as diff\n",
    "# import scDART.dataset as dataset\n",
    "# import scDART.model as model\n",
    "# import scDART.loss as loss\n",
    "# import scDART.train as train\n",
    "# import scDART.post_align as palign\n",
    "# import scDART.benchmark as bmk\n",
    "\n",
    "# # step by step\n",
    "# seeds = [0]\n",
    "# latent_dim = 4\n",
    "# learning_rate = 3e-4\n",
    "# n_epochs = 500\n",
    "# use_anchor = False\n",
    "# reg_d = 1\n",
    "# reg_g = 1\n",
    "# reg_mmd = 1\n",
    "# ts = [30, 50, 70]\n",
    "# use_potential = True\n",
    "\n",
    "# seed = seeds[0]\n",
    "# print(\"Random seed: \" + str(seed))\n",
    "# torch.manual_seed(seed)\n",
    "# torch.cuda.manual_seed(seed)\n",
    "# np.random.seed(seed)\n",
    "\n",
    "# counts_rna = pd.read_csv(\"../data/Endo/counts_rna.csv\", index_col = 0)\n",
    "# counts_atac = pd.read_csv(\"../data/Endo/counts_atac.csv\", index_col = 0)\n",
    "# label_rna = pd.read_csv(\"../data/Endo/anno_rna.txt\", header = None)\n",
    "# label_atac = pd.read_csv(\"../data/Endo/anno_atac.txt\", header = None)\n",
    "# rna_dataset = dataset.dataset(counts = counts_rna.values, anchor = None)\n",
    "# atac_dataset = dataset.dataset(counts = counts_atac.values, anchor = None)\n",
    "# coarse_reg = torch.FloatTensor(pd.read_csv(\"../data/Endo/region2gene.csv\", sep = \",\", index_col = 0).values).to(device)\n",
    "\n",
    "# batch_size = int(max([len(rna_dataset),len(atac_dataset)])/4)\n",
    "\n",
    "# train_rna_loader = DataLoader(rna_dataset, batch_size = batch_size, shuffle = True)\n",
    "# train_atac_loader = DataLoader(atac_dataset, batch_size = batch_size, shuffle = True)\n",
    "\n",
    "\n",
    "# EMBED_CONFIG = {\n",
    "#     'gact_layers': [atac_dataset.counts.shape[1], 1024, 512, rna_dataset.counts.shape[1]], \n",
    "#     'proj_layers': [rna_dataset.counts.shape[1], 512, 128, latent_dim], # number of nodes in each \n",
    "#     'learning_rate': learning_rate,\n",
    "#     'n_epochs': n_epochs + 1,\n",
    "#     'use_anchor': use_anchor,\n",
    "#     'reg_d': reg_d,\n",
    "#     'reg_g': reg_g,\n",
    "#     'reg_mmd': reg_mmd,\n",
    "#     'l_dist_type': 'kl',\n",
    "#     'device': device\n",
    "# }\n",
    "\n",
    "\n",
    "# # calculate the diffusion distance\n",
    "# dist_rna = diff.diffu_distance(rna_dataset.counts.numpy(), ts = ts,\n",
    "#                                 use_potential = use_potential, dr = \"pca\", n_components = 30)\n",
    "\n",
    "# dist_atac = diff.diffu_distance(atac_dataset.counts.numpy(), ts = ts,\n",
    "#                                 use_potential = use_potential, dr = \"lsi\", n_components = 30)\n",
    "\n",
    "\n",
    "# dist_rna = dist_rna/np.linalg.norm(dist_rna)\n",
    "# dist_atac = dist_atac/np.linalg.norm(dist_atac)\n",
    "# dist_rna = torch.FloatTensor(dist_rna).to(device)\n",
    "# dist_atac = torch.FloatTensor(dist_atac).to(device)\n",
    "\n",
    "# # initialize the model\n",
    "# gene_act = model.gene_act(features = EMBED_CONFIG[\"gact_layers\"], dropout_rate = 0.0, negative_slope = 0.2).to(device)\n",
    "# encoder = model.Encoder(features = EMBED_CONFIG[\"proj_layers\"], dropout_rate = 0.0, negative_slope = 0.2).to(device)\n",
    "# model_dict = {\"gene_act\": gene_act, \"encoder\": encoder}\n",
    "\n",
    "# opt_genact = torch.optim.Adam(gene_act.parameters(), lr = learning_rate)\n",
    "# opt_encoder = torch.optim.Adam(encoder.parameters(), lr = learning_rate)\n",
    "# opt_dict = {\"gene_act\": opt_genact, \"encoder\": opt_encoder}\n",
    "\n",
    "# # training models\n",
    "# train.match_latent(model = model_dict, opts = opt_dict, dist_atac = dist_atac, dist_rna = dist_rna, \n",
    "#                 data_loader_rna = train_rna_loader, data_loader_atac = train_atac_loader, n_epochs = EMBED_CONFIG[\"n_epochs\"], \n",
    "#                 reg_mtx = coarse_reg, reg_d = EMBED_CONFIG[\"reg_d\"], reg_g = EMBED_CONFIG[\"reg_g\"], reg_mmd = EMBED_CONFIG[\"reg_mmd\"], use_anchor = EMBED_CONFIG[\"use_anchor\"], norm = \"l1\", \n",
    "#                 mode = EMBED_CONFIG[\"l_dist_type\"])\n",
    "\n",
    "# with torch.no_grad():\n",
    "#     z_rna = model_dict[\"encoder\"](rna_dataset.counts.to(device)).cpu().detach()\n",
    "#     z_atac = model_dict[\"encoder\"](model_dict[\"gene_act\"](atac_dataset.counts.to(device))).cpu().detach()\n",
    "\n",
    "\n",
    "# with torch.no_grad():\n",
    "#     z_rna, z_atac = palign.match_alignment(z_rna = z_rna, z_atac = z_atac, k = 10)\n",
    "#     z_atac, z_rna = palign.match_alignment(z_rna = z_atac, z_atac = z_rna, k = 10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# post-maching\n",
    "pca_op = PCA(n_components = 2)\n",
    "z = pca_op.fit_transform(np.concatenate((z_rna, z_atac), axis = 0))\n",
    "z_rna_pca = z[:z_rna.shape[0],:]\n",
    "z_atac_pca = z[z_rna.shape[0]:,:]\n",
    "_ = utils.plot_latent(z1 = z_rna_pca, z2 = z_atac_pca, anno1 = label_rna, \n",
    "                    anno2 = label_atac, mode = \"joint\", save = None, \n",
    "                    figsize = (15,7), axis_label = \"PCA\")\n",
    "_ = utils.plot_latent(z1 = z_rna_pca, z2 = z_atac_pca, anno1 = label_rna, \n",
    "                    anno2 = label_atac, mode = \"modality\", save = None, \n",
    "                    figsize = (15,7), axis_label = \"PCA\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Infer trajectory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "# define the root cell\n",
    "root_cell = 35\n",
    "\n",
    "# infer the trajectory backbone \n",
    "groups, mean_cluster, T = ti.backbone_inf(np.concatenate((z_rna, z_atac), axis = 0), resolution = 0.05)\n",
    "groups_rna = groups[:counts_rna.shape[0]]\n",
    "groups_atac = groups[counts_rna.shape[0]:]\n",
    "root_clust = groups[root_cell]\n",
    "\n",
    "G = nx.from_numpy_matrix(T, create_using=nx.DiGraph)\n",
    "G = nx.dfs_tree(G, source = root_clust)\n",
    "paths = []\n",
    "for node in G:\n",
    "    if G.out_degree(node)==0: #it's a leaf\n",
    "        paths.append(nx.shortest_path(G, root_clust, node))\n",
    "        \n",
    "pca_op = PCA(n_components = 2)\n",
    "z = pca_op.fit_transform(np.concatenate((z_rna, z_atac), axis = 0))\n",
    "z_rna_pca = z[:z_rna.shape[0],:]\n",
    "z_atac_pca = z[z_rna.shape[0]:,:]    \n",
    "mean_cluster = pca_op.transform(np.array(mean_cluster))\n",
    "utils.plot_backbone(z_rna_pca, z_atac_pca, groups = groups, T = T, mean_cluster = mean_cluster, mode = \"joint\", figsize=(10,7), save = None, axis_label = \"PCA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infer pseudo-time\n",
    "dpt_mtx = ti.dpt(np.concatenate((z_rna, z_atac), axis = 0), n_neigh = 10)\n",
    "pt_infer = dpt_mtx[root_cell, :]\n",
    "pt_infer[pt_infer.argsort()] = np.arange(len(pt_infer))\n",
    "pt_infer = pt_infer/np.max(pt_infer)\n",
    "# for scRNA-Seq batch\n",
    "pt_infer_rna = pt_infer[:z_rna.shape[0]]\n",
    "# for scATAC-Seq batch\n",
    "pt_infer_atac = pt_infer[z_rna.shape[0]:]\n",
    "utils.plot_latent_pt(z1 = z_rna_pca, z2 = z_atac_pca, pt1 = pt_infer_rna, pt2 = pt_infer_atac, mode = \"joint\", save = None, figsize = (10,7), axis_label = \"PCA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a03228faed9f66c6b441de307d12f03baa2532247873a44ca1a954025fabf88d"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
